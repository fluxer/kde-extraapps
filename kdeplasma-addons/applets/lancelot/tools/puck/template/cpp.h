/*
 *   This file is automatically generated by Plasma UI Compiler
 *   DO NOT CHANGE THE CONTENTS, IT IS AUTOMATICALLY REGENERATED
 */

#ifndef ${HEADER_ID}
#define ${HEADER_ID}

/* Includes */
${INCLUDES}

/* Defines */
${DEFINES}

namespace Ui {

class ${CLASS_NAME}
{

protected:

    ${DECLARATION}
    
    void setupUi(${PARENT_OBJECT_TYPE} * root)
    {
        /* Initialization */
        ${INITIALIZATION}
        
        /* Setup */
        ${SETUP}
    }
};

} // namespace Ui

#endif /* ${HEADER_ID} */
