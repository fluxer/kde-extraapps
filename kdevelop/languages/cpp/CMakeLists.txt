project(cpp)
include(CppBuildConfig.cmake)

include_directories(
    ${KDevelop_SOURCE_DIR}/languages/cpp/parser
    ${KDevelop_SOURCE_DIR}/languages/cpp/cppduchain
)

add_definitions( -DKDE_DEFAULT_DEBUG_AREA=9007 )

add_subdirectory(parser)
add_subdirectory(cppduchain)
add_subdirectory(tests)
########### next target ###############

set(kdevcpplanguagesupport_PART_SRCS
    cpplanguagesupport.cpp
    includepathcomputer.cpp
    cppparsejob.cpp
    preprocessjob.cpp
    cpphighlighting.cpp
    cpputils.cpp
    includepathresolver.cpp
    setuphelpers.cpp
    quickopen.cpp
    
    codecompletion/model.cpp
    codecompletion/worker.cpp
    codecompletion/context.cpp
    codecompletion/item.cpp
    codecompletion/helpers.cpp
    codecompletion/missingincludeitem.cpp
    codecompletion/implementationhelperitem.cpp

#    codegen/cppnewclass.cpp
    codegen/simplerefactoring.cpp
    codegen/adaptsignatureassistant.cpp
    codegen/adaptsignatureaction.cpp
    codegen/unresolvedincludeassistant.cpp
    codegen/customincludepaths.cpp
    codegen/cppclasshelper.cpp
#     codegen/makeimplementationprivate.cpp
)

set(kdevcpplanguagesupport_LIB_UI
    codegen/ui/custom_include_paths.ui
#     codegen/ui/privateimplementation.ui
)

kde4_add_ui_files(kdevcpplanguagesupport_PART_SRCS ${kdevcpplanguagesupport_LIB_UI})

option(BUILD_kdev_includepathresolver "Build the includepath resolver debugging tool" OFF)
if(BUILD_kdev_includepathresolver)
  add_executable(kdev_includepathresolver includepathresolver.cpp)
  set_target_properties( kdev_includepathresolver PROPERTIES COMPILE_FLAGS -DTEST )
  target_link_libraries( kdev_includepathresolver
  ${KDEVPLATFORM_INTERFACES_LIBRARIES}  ${KDEVPLATFORM_PROJECT_LIBRARIES}
  ${QT_QTCORE_LIBRARY} ${QT_QTGUI_LIBRARY} ${KDE4_KDECORE_LIBS} ${KDEVPLATFORM_LANGUAGE_LIBRARIES} )
  install(TARGETS kdev_includepathresolver ${INSTALL_TARGETS_DEFAULT_ARGS} )
endif(BUILD_kdev_includepathresolver)

option(BUILD_uiblocktester "Build a debug thread into the plugin that checks for UI-thread lockups" OFF)
if(BUILD_uiblocktester)
  add_definitions(-DDEBUG_UI_LOCKUP)
endif(BUILD_uiblocktester)

kde4_add_plugin(kdevcpplanguagesupport ${kdevcpplanguagesupport_PART_SRCS})
target_link_libraries(kdevcpplanguagesupport
    kdev4cpprpp
    kdev4cppduchain
    kdev4cppparser
    ${KDEVPLATFORM_INTERFACES_LIBRARIES}
    ${KDEVPLATFORM_LANGUAGE_LIBRARIES}
    ${KDEVPLATFORM_PROJECT_LIBRARIES}
    ${KDEVPLATFORM_UTIL_LIBRARIES}
    ${KDE4_THREADWEAVER_LIBRARIES}
    ${KDE4_KDECORE_LIBS}
    ${KDE4_KTEXTEDITOR_LIBS}
)

install(TARGETS kdevcpplanguagesupport DESTINATION ${PLUGIN_INSTALL_DIR})


########### install files ###############

configure_file(kdevcppsupport.desktop.cmake ${CMAKE_CURRENT_BINARY_DIR}/kdevcppsupport.desktop)
install(FILES ${CMAKE_CURRENT_BINARY_DIR}/kdevcppsupport.desktop DESTINATION ${SERVICES_INSTALL_DIR})
install(FILES kdevcppsupport.rc DESTINATION ${DATA_INSTALL_DIR}/kdevcppsupport)
