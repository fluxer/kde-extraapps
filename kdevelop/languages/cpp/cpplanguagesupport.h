/*
 * KDevelop C++ Language Support
 *
 * Copyright 2005 Matt Rogers <mattr@kde.org>
 * Copyright 2007-2008 David Nolden<david.nolden.kdevelop@art-master.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Library General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#ifndef KDEVCPPLANGUAGESUPPORT_H
#define KDEVCPPLANGUAGESUPPORT_H

#include <interfaces/iplugin.h>
#include <language/interfaces/ilanguagesupport.h>
#include <language/duchain/topducontext.h>
#include <interfaces/ibuddydocumentfinder.h>

#include <QDateTime>
#include <QThread>
#include <QTimer>
#include <QMutex>

namespace rpp {
class pp_macro;
}

class CppHighlighting;
class CPPParseJob;
class IncludeFileDataProvider;
class SimpleRefactoring;

namespace KDevelop {
  class ICodeHighlighting;
  class SimpleRange;
  class CodeCompletion;
}
namespace Cpp {
  class StaticCodeAssistant;
}
namespace CppTools {
  class IncludePathResolver;
}

///A class that helps detecting what exactly makes the UI block. To use it, just place a breakpoint on UIBlockTester::lockup() and inspect the execution-position of the main thread
class UIBlockTester : public QObject {
Q_OBJECT
  class UIBlockTesterThread : public QThread {
    public:
      UIBlockTesterThread( UIBlockTester& parent );
      void run();
      void stop();
    private:
      UIBlockTester& m_parent;
      bool m_stop;
  };
  friend class UIBlockTesterThread;
public:

  ///@param milliseconds when the ui locks for .. milliseconds, lockup() is called
  UIBlockTester( uint milliseconds, QObject* parent );
  ~UIBlockTester();

private slots:
  void timer();

protected:
   virtual void lockup();

 private:
     UIBlockTesterThread m_thread;
     QDateTime m_lastTime;
     QMutex m_timeMutex;
     QTimer * m_timer;
     uint m_msecs;
};

class CppLanguageSupport : public KDevelop::IPlugin, public KDevelop::ILanguageSupport,
                           public KDevelop::IBuddyDocumentFinder
{
Q_OBJECT
Q_INTERFACES( KDevelop::ILanguageSupport )

public:
    explicit CppLanguageSupport( QObject* parent, const QVariantList& args = QVariantList() );
    virtual ~CppLanguageSupport();

    QString name() const;

    KDevelop::ICodeHighlighting *codeHighlighting() const;
    KDevelop::BasicRefactoring* refactoring() const;
    KDevelop::ILanguage *language();
    KDevelop::ContextMenuExtension contextMenuExtension(KDevelop::Context* context);
    KDevelop::ParseJob *createParseJob( const KDevelop::IndexedString &url );
    //KDevelop::AstRepresentationPtr  generateAst(const KDevelop::TopDUContext & topContext);
    KDevelop::ICreateClassHelper* createClassHelper() const;
    
    static CppLanguageSupport* self();

    virtual QString indentationSample() const {
      return "class C{\n class D {\n void c() {\n int m;\n }\n }\n};\n";
    }
    
    virtual void createActionsForMainWindow(Sublime::MainWindow* window, QString& xmlFile, KActionCollection& actions);
/**
 * There may be multiple differnt parsed versions of a document available in the du-chain.
 * This function helps choosing the right one, by creating a standard parsing-environment,
 * and searching for a TopDUContext that fits in. If this fails, a random version is chosen.
 *
 * If simplified environment-matching is enabled, and a proxy-context is found, it returns
 * that proxy-contexts target-context, so the returned context may be used for completion etc.
 * without additional checking.
 *
 * @todo Move this somewhere more general
 *
 * @warning The du-chain must be locked before calling this.
* */
  virtual KDevelop::TopDUContext *standardContext(const KUrl& url, bool proxyContext = false);

    /**
     * IBuddyDocumentFinder overrides.
     */
    virtual bool areBuddies(const KUrl& url1, const KUrl& url2);
    virtual bool buddyOrder(const KUrl& url1, const KUrl& url2);
    virtual QVector<KUrl> getPotentialBuddies(const KUrl& url) const;

public slots:
    void findIncludePathsForJob(CPPParseJob* job);

    ///UI:
    void switchDefinitionDeclaration();

private:

    //Returns the identifier and its range under the cursor as first return-value, and the tail behind it as the second
    //If the given line is an include directive, the complete line is returned starting at the directive
    QPair<QPair<QString, KDevelop::SimpleRange>, QString> cursorIdentifier(const KUrl& url, const KDevelop::SimpleCursor& position) const;

    QPair<KDevelop::TopDUContextPointer, KDevelop::SimpleRange> importedContextForPosition(const KUrl& url, const KDevelop::SimpleCursor& position);

    QPair<KDevelop::SimpleRange, const rpp::pp_macro*> usedMacroForPosition(const KUrl& url, const KDevelop::SimpleCursor& position);

    virtual KDevelop::SimpleRange specialLanguageObjectRange(const KUrl& url, const KDevelop::SimpleCursor& position);

    virtual QPair<KUrl, KDevelop::SimpleCursor> specialLanguageObjectJumpCursor(const KUrl& url, const KDevelop::SimpleCursor& position);

    virtual QWidget* specialLanguageObjectNavigationWidget(const KUrl& url, const KDevelop::SimpleCursor& position);

    static CppLanguageSupport* m_self;

    CppHighlighting *m_highlights;
    SimpleRefactoring *m_refactoring;
    KDevelop::CodeCompletion *m_cc;

    IncludeFileDataProvider* m_quickOpenDataProvider;
    
    const QStringList m_mimeTypes;
};

#endif

