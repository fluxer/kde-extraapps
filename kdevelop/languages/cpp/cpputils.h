/*
   Copyright 2007-2008 David Nolden <david.nolden.kdevelop@art-master.de>

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License version 2 as published by the Free Software Foundation.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public License
   along with this library; see the file COPYING.LIB.  If not, write to
   the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
   Boston, MA 02110-1301, USA.
*/

#ifndef CPPUTILS_H
#define CPPUTILS_H

#include "environmentmanager.h"
#include <util/path.h>

namespace KDevelop
{
class IncludeItem;
class Declaration;
class ParsingEnvironment;
}

class KUrl;

namespace CppUtils
{
  
///Returns -1 if there is no #include in this line, else the first position behind the "#include" string, which may also look like " # include" or similar
int findEndOfInclude(QString line);

///If @param fast is true, no exhaustive search is done as fallback.
QString sourceOrHeaderCandidate( const QString &path, bool fast = false );

///Returns true if the given url is a header, looking at he known file extensions
bool isHeader(const KUrl &url);

const Cpp::ReferenceCountedMacroSet& standardMacros();

/// Get the full path for a file based on a search through the project's
/// include directories
/// @param localPath the path from which this findInclude is issued
/// @param skipPath this path will be skipped while searching, as needed for gcc extension #include_next
/// @return first: The found file(not a canonical path, starts with the directory it was found in)
///         second: The include-path the file was found in(can be used to skip that path on #include_next)
QPair<KDevelop::Path, KDevelop::Path> findInclude(const KDevelop::Path::List& includePaths, const KDevelop::Path& localPath,
                                                  const QString& includeName, int includeType, const KDevelop::Path& skipPath,
                                                  bool quiet=false);
    
///Thread-safe
bool needsUpdate(const Cpp::EnvironmentFilePointer& file, const KDevelop::Path& localPath, const KDevelop::Path::List& includePaths );

///Returns the include-path. Each dir has a trailing slash. Search should be iterated forward through the list
// NOTE: _no_ lock must be held when calling this, otherwise you might deadlock since this is a very ugly hacky function
// which must call stuff from the foreground thread...
KDevelop::Path::List findIncludePaths(const QString& source);

/**
  * Returns a list of all files within the include-path of the given file
  * @param addPath This path is added behind each include-path, and the content of the resulting directory used.
  * @param addIncludePaths A list of include-paths that should be used for listing, additionally to the known ones
  * @param onlyAddedIncludePaths If this is true, only the include-paths given in @p addIncludePaths will be used
  * @param prependAddedPathToName If this is true, @p addPath is prepended to each of the returned items paths
  * */
QList<KDevelop::IncludeItem> allFilesInIncludePath(const QString& source, bool local, const QString& addPath,
                                                   const QStringList& addIncludePaths = QStringList(), bool onlyAddedIncludePaths = false,
                                                   bool prependAddedPathToName = false, bool allowSourceFiles = false );

/// @return List of possible header extensions used for definition/declaration fallback switching
QStringList headerExtensions();
/// @return List of possible source extensions used for definition/declaration fallback switching
QStringList sourceExtensions();

}


#endif	//CPPUTILS_H
