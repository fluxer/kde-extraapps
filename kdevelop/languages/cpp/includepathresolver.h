/***************************************************************************
   Copyright 2006 David Nolden <david.nolden.kdevelop@art-master.de>
***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
#ifndef INCLUDEPATHRESOLVER_H
#define INCLUDEPATHRESOLVER_H

#include <QString>
#include <QStringList>
#include <language/editor/modificationrevisionset.h>

class KUrl;
class QDir;
class QFile;

namespace CppTools {

class FileModificationTimeWrapper;

struct PathResolutionResult
{
  PathResolutionResult(bool success = false, const QString& errorMessage = QString(), const QString& longErrorMessage = QString())
    : success(success)
    , errorMessage(errorMessage)
    , longErrorMessage(longErrorMessage)
  {
  }
  bool success;
  QString errorMessage;
  QString longErrorMessage;

  KDevelop::ModificationRevisionSet includePathDependency;

  QStringList paths;

  void addPathsUnique(const PathResolutionResult& rhs)
  {
    foreach(const QString& path, rhs.paths) {
      if(!paths.contains(path))
        paths.append(path);
    }
    includePathDependency += rhs.includePathDependency;
  }

  operator bool() const
  {
    return success;
  }
};

struct CustomIncludePathsSettings
{
  QString storagePath; //Directory the custom file is in
  QString sourceDir;
  QString buildDir;
  QStringList paths;

  QString storageFile() const;
  bool isValid() const ;

  static CustomIncludePathsSettings read(const QString& storagePath);
  ///Finds a valid storage file above the given start path that contains custom include paht settings
  ///If no valid storage file is found, returns an empty string
  static QString find(const QString& startPath);
  ///Finds a storage-path for the given start path, and reads the custom include path settings
  ///If none were found, returns an invalid item
  static CustomIncludePathsSettings findAndRead(const QString& startPath);
  ///Same as findAndRead, but also convert all relative paths into absolute ones from the
  ///storage path.
  static CustomIncludePathsSettings findAndReadAbsolute(const QString& startPath);

  //Stores these settings exclusively, overwriting any old ones for the storage path
  bool write();
  //Deletes these settings
  bool delete_();
};

class SourcePathInformation;

///One resolution-try can issue up to 4 make-calls in worst case
class IncludePathResolver
{
  public:
    IncludePathResolver();
    ///Same as below, but uses the directory of the file as working-directory. The argument must be absolute.
    PathResolutionResult resolveIncludePath( const QString& file );
    ///The include-path is only computed once for a whole directory, then it is cached using the modification-time of the Makefile.
    PathResolutionResult resolveIncludePath( const QString& file, const QString& workingDirectory, int maxStepsUp = 20 );
    ///source and build must be absolute paths
    void setOutOfSourceBuildSystem( const QString& source, const QString& build );
    ///resets to in-source build system
    void resetOutOfSourceBuild();

    static void clearCache();

    KDevelop::ModificationRevisionSet findIncludePathDependency(const QString& file);

    void enableMakeResolution(bool enable);

  private:
    bool m_isResolving;
    bool m_outOfSource;
    bool m_enableMakeResolution;

    KUrl mapToBuild(const KUrl& url);

    ///Executes the command using KProcess
    bool executeCommand( const QString& command, const QString& workingDirectory, QString& result ) const;
    ///file should be the name of the target, without extension(because that may be different)
    PathResolutionResult resolveIncludePathInternal( const QString& file, const QString& workingDirectory,
                                                      const QString& makeParameters, const SourcePathInformation& source, int maxDepth );
    QString m_source;
    QString m_build;
};

}

#endif
