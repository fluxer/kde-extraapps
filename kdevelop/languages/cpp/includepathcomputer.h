/*
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License version 2 as published by the Free Software Foundation.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public License
   along with this library; see the file COPYING.LIB.  If not, write to
   the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
   Boston, MA 02110-1301, USA.
*/

#ifndef INCLUDEPATHCOMPUTER_H
#define INCLUDEPATHCOMPUTER_H

#include <util/path.h>

#include "includepathresolver.h"

class IncludePathComputer
{
public:
  IncludePathComputer(const QString& file);
  ///Must be called in the foreground thread, before calling computeBackground().
  void computeForeground();
  ///Can be called from within background thread, but does not have to. May lock for a long time.
  void computeBackground();

  KDevelop::Path::List result() const
  {
    return m_ret;
  }

  QHash<QString,QString> defines() const
  {
    return m_defines;
  }

  KDevelop::ModificationRevisionSet m_includePathDependency;

private:
  void addInclude(const KDevelop::Path& path);

  QHash<QString,QString> m_defines;
  QString m_source;
  KDevelop::Path::List m_ret;
  QSet<KDevelop::Path> m_hasPath;
  bool m_ready;

  KDevelop::Path m_effectiveBuildDirectory;
  KDevelop::Path m_buildDirectory;
  KDevelop::Path m_projectDirectory;
  QString m_projectName;
  bool m_gotPathsFromManager;
  CppTools::IncludePathResolver m_includeResolver;
};
#endif // INCLUDEPATHCOMPUTER_H
