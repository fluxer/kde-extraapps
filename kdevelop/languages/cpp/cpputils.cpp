/*
   Copyright 2007-2008 David Nolden <david.nolden.kdevelop@art-master.de>

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License version 2 as published by the Free Software Foundation.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public License
   along with this library; see the file COPYING.LIB.  If not, write to
   the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
   Boston, MA 02110-1301, USA.
*/

#include "cpputils.h"

#include "setuphelpers.h"
#include "parser/rpp/preprocessor.h"
#include "includepathcomputer.h"

#include <interfaces/icore.h>
#include <interfaces/iprojectcontroller.h>
#include <interfaces/iproject.h>

#include <language/codegen/coderepresentation.h>
#include <language/duchain/declaration.h>
#include <language/duchain/duchain.h>
#include <language/duchain/duchainlock.h>
#include <language/util/includeitem.h>

#include "interfaces/foregroundlock.h"

#include <project/projectmodel.h>

#include <QDirIterator>
#include <QThread>
#include <QCoreApplication>

template<class T>
static QList<T> makeListUnique(const QList<T>& list)
{
  QList<T> ret;

  QSet<QString> set;
  foreach(const T& item, list)
  {
    if(!set.contains(item))
    {
      ret << item;
      set.insert(item);
    }
  }
  
  return ret;
}

QString addDot(QString ext) {
  if(ext.contains('.')) //We need this check because of the _impl.h thing
    return ext;
  else
    return "." + ext;
}

using namespace KDevelop;

namespace CppUtils
{
  
int findEndOfInclude(QString line)
{
  QString tmp = line;
  tmp = tmp.trimmed();
  if(!tmp.startsWith("#"))
    return -1;
  
  tmp = tmp.mid(1).trimmed();

  if(!tmp.startsWith("include"))
    return -1;
  
  return line.indexOf("include") + 7;
}

QString sourceOrHeaderCandidate( const QString &path_, bool fast )
{
// get the path of the currently active document
  QFileInfo fi( path_ );
  QString path = fi.filePath();
  // extract the exension
  QString ext = fi.suffix();
  if ( ext.isEmpty() )
    return QString();
  // extract the base path (full path without '.' and extension)
  QString base = path.left( path.length() - ext.length() - 1 );
  //kDebug( 9007 ) << "base: " << base << ", ext: " << ext << endl;
  // just the filename without the extension
  QString fileNameWoExt = fi.fileName();
  if ( !ext.isEmpty() )
    fileNameWoExt.replace( "." + ext, "" );
  QStringList possibleExts;
  // depending on the current extension assemble a list of
  // candidate files to look for
  QStringList candidates;
  // special case for template classes created by the new class dialog
  if ( path.endsWith( "_impl.h" ) )
  {
    QString headerpath = path;
    headerpath.replace( "_impl.h", ".h" );
    candidates << headerpath;
    fileNameWoExt.replace( "_impl", "" );
    possibleExts << "h";
  }
  // if file is a header file search for implementation file
  else if ( headerExtensions().contains( ext ) )
  {
    foreach(const QString& ext, sourceExtensions())
      candidates << ( base + addDot(ext) );

    possibleExts = sourceExtensions();
  }
  // if file is an implementation file, search for header file
  else if ( sourceExtensions().contains( ext ) )
  {
    foreach(const QString& ext, headerExtensions())
      candidates << ( base + addDot(ext) );

    possibleExts = headerExtensions();
  }
  // search for files from the assembled candidate lists, return the first
  // candidate file that actually exists or QString::null if nothing is found.
  QStringList::ConstIterator it;
  for ( it = candidates.constBegin(); it != candidates.constEnd(); ++it )
  {
//     kDebug( 9007 ) << "Trying " << ( *it ) << endl;
    if ( QFileInfo( *it ).exists() )
    {
//       kDebug( 9007 ) << "using: " << *it << endl;
      return * it;
    }
  }

  if(fast)
    return QString();

  //kDebug( 9007 ) << "Now searching in project files." << endl;
  // Our last resort: search the project file list for matching files

  QFileInfo candidateFileWoExt;
  QString candidateFileWoExtString;

  const IndexedString file(path);
  foreach (KDevelop::IProject *project, ICore::self()->projectController()->projects()) {
      if (project->inProject(file)) {
        foreach(const IndexedString& otherFile, project->fileSet()) {
          candidateFileWoExt.setFile(otherFile.str());
          //kDebug( 9007 ) << "candidate file: " << otherFile.str() << endl;
          if( !candidateFileWoExt.suffix().isEmpty() )
            candidateFileWoExtString = candidateFileWoExt.fileName().replace( "." + candidateFileWoExt.suffix(), "" );

          if ( candidateFileWoExtString == fileNameWoExt )
          {
            if ( possibleExts.contains( candidateFileWoExt.suffix() ) || candidateFileWoExt.suffix().isEmpty() )
            {
              //kDebug( 9007 ) << "checking if " << url << " exists" << endl;
              return otherFile.str();
            }
          }
        }
      }
  }

  return QString();
}


bool isHeader(const KUrl &url) {
  QFileInfo fi( url.toLocalFile() );
  QString path = fi.filePath();
  // extract the exension
  QString ext = fi.suffix();
  if ( ext.isEmpty() )
    return true;
  
  return headerExtensions().contains(ext);
}

const Cpp::ReferenceCountedMacroSet& standardMacros()
{
  static Cpp::ReferenceCountedMacroSet macros = CppTools::setupStandardMacros();
  return macros;
}

QPair<Path, Path> findInclude(const Path::List& includePaths, const Path& localPath,
                              const QString& includeName, int includeType,
                              const Path& skipPath, bool quiet){
    QPair<Path, Path> ret;
#ifdef DEBUG
    kDebug(9007) << "searching for include-file" << includeName;
    if( !skipPath.isEmpty() )
        kDebug(9007) << "skipping path" << skipPath;
#endif

    if (includeName.startsWith('/')) {
        QFileInfo info(includeName);
        if (info.exists() && info.isReadable() && info.isFile()) {
            //kDebug(9007) << "found include file:" << info.absoluteFilePath();
            ret.first = Path(info.canonicalFilePath());
            ret.second = Path("/");
            return ret;
        }
    }

    if (includeType == rpp::Preprocessor::IncludeLocal && localPath != skipPath) {
        Path check(localPath, includeName);
        QFileInfo info(check.toLocalFile());
        if (info.exists() && info.isReadable() && info.isFile()) {
            //kDebug(9007) << "found include file:" << info.absoluteFilePath();
            ret.first = check;
            ret.second = localPath;
            return ret;
        }
    }

    //When a path is skipped, we will start searching exactly after that path
    bool needSkip = skipPath.isValid();

restart:
    foreach( const Path& path, includePaths ) {
        if( needSkip ) {
            if( path == skipPath ) {
                needSkip = false;
                continue;
            }
        }

        Path check(path, includeName);
        QFileInfo info(check.toLocalFile());

        if (info.exists() && info.isReadable() && info.isFile()) {
            //kDebug(9007) << "found include file:" << info.absoluteFilePath();
            ret.first = check;
            ret.second = path;
            return ret;
        }
    }

    if( needSkip ) {
        //The path to be skipped was not found, so simply start from the begin, considering any path.
        needSkip = false;
        goto restart;
    }

    const int idx = includeName.indexOf('/');
    if ( idx != -1 ) {
      // HACK: parse Qt4 includes and similar even without the full include paths from the project manager
      // there, a file in /usr/include/qt4/QtCore/ tries to include sibling files via QtCore/file
      ret = findInclude(includePaths, localPath, includeName.mid(idx + 1), rpp::Preprocessor::IncludeLocal, skipPath, false);
    }

    if( !ret.first.isValid())
    {
        //Check if there is an available artificial representation
        if(!includeName.isNull() && artificialCodeRepresentationExists(IndexedString(includeName)))
        {
            ret.first = Path(CodeRepresentation::artificialPath(includeName));
            kDebug() << "Utilizing Artificial code for include: " << includeName;
        }
        else if(!quiet ) {
            kDebug() << "FAILED to find include-file" << includeName << "in paths:" << includePaths;
        }
    }
    
    return ret;
}

bool needsUpdate(const Cpp::EnvironmentFilePointer& file, const Path& localPath, const Path::List& includePaths)
{
  if(file->needsUpdate())
    return true;

  ///@todo somehow this check should also go into EnvironmentManager
  for( Cpp::ReferenceCountedStringSet::Iterator it = file->missingIncludeFiles().iterator(); it; ++it ) {

    ///@todo store IncludeLocal/IncludeGlobal somewhere
    auto included = findInclude( includePaths, localPath, (*it).str(), rpp::Preprocessor::IncludeLocal, Path(), true );
    if(included.first.isValid())
      return true;
  }

  return false;
}

/**
 * This is an ugly HACK but well... it works and doesn't require an extremely big API refactoring.
 *
 * Thing is, the findIncludePaths was potentially called from background threads, mostly during
 * code completion. From there, it is not safe to call IncludePathComputer::computeForeground.
 * To fix this, we now add another eventloop roundtrip, execute that function in the foreground
 * and then afterwards compute the rest in the background. This does mean that the background thread
 * is _blocked_ while we compute stuff in the foreground.
 *
 * This must be properly fixed when we port the MissingInclude feature to clang. There, the
 * include paths must be queried in the foreground thread, before invoking code completion.
 */
class IncludePathForegroundComputer : public DoInForeground
{
public:
    IncludePathForegroundComputer(IncludePathComputer* includePathComputer)
      : m_includePathComputer(includePathComputer)
    {}

private:
    void doInternal() override final
    {
      m_includePathComputer->computeForeground();
    }

    IncludePathComputer* m_includePathComputer;
};

Path::List findIncludePaths(const QString& source)
{
  Q_ASSERT(QThread::currentThread() == qApp->thread() ||
           (!DUChain::lock()->currentThreadHasReadLock() && !DUChain::lock()->currentThreadHasWriteLock()));

  IncludePathComputer comp(source);
  IncludePathForegroundComputer foreground(&comp);
  foreground.doIt();
  comp.computeBackground();
  return comp.result();
}

QList<KDevelop::IncludeItem> allFilesInIncludePath(const QString& source, bool local, const QString& addPath,
                                                   const QStringList& addIncludePaths, bool onlyAddedIncludePaths,
                                                   bool prependAddedPathToName, bool allowSourceFiles)
{
    QList<KDevelop::IncludeItem> ret;

    QStringList paths;
    if ( addPath.startsWith('/') ) {
      paths << QString("/");
    } else {
      paths = addIncludePaths;
      if(!onlyAddedIncludePaths) {
        foreach(const Path& path, findIncludePaths(source)) {
          paths += path.toLocalFile();
        }

        if(local) {
            KUrl localPath(source);
            localPath.setFileName(QString());
            paths.push_front(localPath.toLocalFile());
        }
      }
    }

    paths = makeListUnique<QString>(paths);
    int pathNumber = 0;

    QSet<QString> hadIncludePaths;
    foreach(const QString& path, paths)
    {
        if(hadIncludePaths.contains(path)) {
          continue;
        }
        hadIncludePaths.insert(path);
        QString searchPath = path;
        if (!addPath.isEmpty() && !addPath.startsWith('/')) {
          if (!searchPath.endsWith('/')) {
            searchPath += '/';
          }
          searchPath += addPath;
        }

        QDirIterator dirContent(searchPath);

        while(dirContent.hasNext()) {
             dirContent.next();
            KDevelop::IncludeItem item;
            item.name = dirContent.fileName();

            if(item.name.startsWith('.') || item.name.endsWith("~")) //This filters out ".", "..", and hidden files, and backups
              continue;
            QString suffix = dirContent.fileInfo().suffix();
            if(!dirContent.fileInfo().suffix().isEmpty() && !headerExtensions().contains(suffix) && (!allowSourceFiles || !sourceExtensions().contains(suffix)))
              continue;
            
            QString fullPath = dirContent.fileInfo().canonicalFilePath();
            if (hadIncludePaths.contains(fullPath)) {
              continue;
            } else {
              hadIncludePaths.insert(fullPath);
            }
            if(prependAddedPathToName) {
              item.name = addPath + item.name;
              item.basePath = path;
            } else {
              item.basePath = searchPath;
            }
            
            item.isDirectory = dirContent.fileInfo().isDir();
            item.pathNumber = pathNumber;

            ret << item;
        }
        ++pathNumber;
    }

    return ret;
}

QStringList headerExtensions()
{
  static const QStringList headerExtensions = QString("h,H,hh,hxx,hpp,tlh,h++").split(',');
  return headerExtensions;
}

QStringList sourceExtensions()
{
  static const QStringList sourceExtensions = QString("c,cc,cpp,c++,cxx,C,m,mm,M,inl,_impl.h").split(',');
  return sourceExtensions;
}

}
