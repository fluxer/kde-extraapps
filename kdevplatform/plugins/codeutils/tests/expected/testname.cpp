/*
 * Test license header
 * In two lines
 */
#include "testname.h"

void TestName::initTestCase()
{
    // Called before the first testfunction is executed
}

void TestName::cleanupTestCase()
{
    // Called after the last testfunction was executed
}

void TestName::init()
{
    // Called before each testfunction is executed
}

void TestName::cleanup()
{
    // Called after every testfunction
}

void TestName::firstTestCase()
{

}

void TestName::secondTestCase()
{

}

void TestName::thirdTestCase()
{

}

QTEST_MAIN(TestName);
#include "moc_testname.cpp"
