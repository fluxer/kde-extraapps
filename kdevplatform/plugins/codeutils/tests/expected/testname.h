/*
 * Test license header
 * In two lines
 */

#ifndef TESTNAME_H
#define TESTNAME_H

#include <QtTest/QtTest>

class TestName : public QObject
{
    Q_OBJECT
private slots:
    void initTestCase();
    void cleanupTestCase();

    void init();
    void cleanup();

    void firstTestCase();
    void secondTestCase();
    void thirdTestCase();
};

#endif // TESTNAME_H
