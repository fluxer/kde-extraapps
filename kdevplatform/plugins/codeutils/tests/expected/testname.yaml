name: TestName
cases:
- name: firstTestCase
- name: secondTestCase
- name: thirdTestCase
